File: content/snippets/cpp/interview_lcci/08/04_power_set.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

vector<vector<int>> subsets(vector<int>& nums) {
    sort(nums.begin(), nums.end());
    int n = nums.size();
    vector<vector<int>> result;
    for (int mask = 0; mask < (1 << n); mask++) {
        vector<int> subset;
        for (int i = 0; i < n; i++) {
            if (mask & (1 << i)) subset.push_back(nums[i]);
        }
        result.push_back(subset);
    }
    return result;
}

int main() {
    vector<int> nums = {1, 2, 3};
    auto result = subsets(nums);
    for (auto& s : result) {
        cout << "[";
        for (int i = 0; i < (int)s.size(); i++) {
            cout << s[i];
            if (i + 1 < (int)s.size()) cout << ",";
        }
        cout << "]" << endl;
    }
    return 0;
}

File: content/snippets/python/interview_lcci/08/04_power_set.py
from typing import List

def subsets(nums: List[int]) -> List[List[int]]:
    nums.sort()
    n = len(nums)
    result = []
    for mask in range(1 << n):
        subset = [nums[i] for i in range(n) if mask & (1 << i)]
        result.append(subset)
    return result

def main():
    for s in subsets([1, 2, 3]):
        print(s)

if __name__ == "__main__":
    main()
